src/components/PaginatedVirtualList/index.tsx

```diff
@@ -0,0 +1,152 @@
+import { useCallback, useEffect, useRef, useState } from "react";
+import { ItemContent, Virtuoso, VirtuosoHandle } from "react-virtuoso";
+import usePrevious from "hooks/usePrevious";
+import { leaveBreadcrumb } from "utils/errorReporting";
+import { calculatePageSize } from "./utils";
+
+interface PaginatedVirtualListProps {
+  count: number;
+  rowRenderer: ItemContent<any, any>;
+  /**
+   * The number of lines to render on each page.
+   */
+  paginationThreshold?: number;
+  /**
+   * The number of lines to scroll by when the user scrolls to the next page.
+   * This is used to avoid the scroll event firing again and causing an infinite loop.
+   * This value must be less than paginationThreshold.
+   */
+  paginationOffset?: number;
+}
+const PaginatedVirtualList: React.FC<PaginatedVirtualListProps> = ({
+  count,
+  rowRenderer,
+  paginationThreshold = 10000,
+  paginationOffset = 10,
+}) => {
+  if (paginationOffset >= paginationThreshold) {
+    throw new Error("paginationOffset must be less than paginationThreshold");
+  }
+  const [currentPage, setCurrentPage] = useState(0);
+  const prevPage = usePrevious(currentPage);
+  const totalPageCount = Math.ceil(count / paginationThreshold);
+  const offsetCompensation = currentPage > 0 ? -paginationOffset : 0;
+
+  const pageSize = calculatePageSize(
+    paginationThreshold,
+    count,
+    currentPage,
+    paginationOffset
+  );
+
+  leaveBreadcrumb(
+    "PaginatedVirtualList",
+    {
+      currentPage,
+      totalPageCount,
+      pageSize,
+    },
+    "process"
+  );
+  const listRef = useRef<VirtuosoHandle>(null);
+
+  const scrollToNextPage = useCallback(() => {
+    const nextPage = currentPage + 1;
+    if (nextPage < totalPageCount) {
+      setCurrentPage(nextPage);
+
+      leaveBreadcrumb(
+        "PaginatedVirtualList",
+        {
+          message: "scrollToNextPage",
+          paginationOffset,
+          paginationThreshold,
+          currentPage,
+          totalPageCount,
+        },
+        "process"
+      );
+    }
+  }, [currentPage, paginationOffset, paginationThreshold, totalPageCount]);
+
+  const scrollToPrevPage = useCallback(() => {
+    const nextPage = currentPage - 1;
+    if (currentPage !== 0) {
+      setCurrentPage(nextPage);
+
+      leaveBreadcrumb(
+        "PaginatedVirtualList",
+        {
+          message: "scrollToPrevPage",
+          offsetCompensation,
+          paginationOffset,
+          paginationThreshold,
+          currentPage,
+          totalPageCount,
+        },
+        "process"
+      );
+    }
+  }, [
+    currentPage,
+    offsetCompensation,
+    paginationOffset,
+    paginationThreshold,
+    totalPageCount,
+  ]);
+
+  useEffect(() => {
+    if (prevPage === undefined) {
+      return;
+    }
+    if (prevPage < currentPage) {
+      listRef.current?.scrollIntoView({ index: paginationOffset });
```

Comment:
I actually couldn't find this `scrollIntoView` function in the docs, but I think we might be able to get a more realistic effect for scrolling to a next page if we used the [`scrollToIndex`](https://virtuoso.dev/scroll-to-index/) function with `align: "end"` behavior. This should make it so that when we scroll past the threshold number, the page shouldn't jump as much

For scrolling to a previous page we would use `align: "start"` I think